kong/core/router.lua

```diff
@@ -128,18 +143,48 @@ local function marshall_api(api)
     if #api.uris > 0 then
       api_t.match_rules = bor(api_t.match_rules, MATCH_RULES.URI)
 
-      for i, uri in ipairs(api.uris) do
-        local escaped_uri = [[\Q]] .. uri .. [[\E]]
-        local strip_regex = escaped_uri .. [[\/?(.*)]]
+      for _, uri in ipairs(api.uris) do
+        if re_match(uri, [[^[a-zA-Z0-9\.\-_~/%]*$]]) then
+          -- plain URI or URI prefix
+          local escaped_uri = [[\Q]] .. uri .. [[\E]]
+          local strip_regex = escaped_uri .. [[/?(?P<stripped_uri>.*)]]
+
+          api_t.uris[uri] = {
+            prefix        = true,
+            strip_regex   = strip_regex,
+          }
+
+          insert(api_t.uris, {
+            prefix      = true,
+            value       = uri,
+            regex       = escaped_uri,
+            strip_regex = strip_regex,
+          })
+
+        else
+          -- regex URI
+          local strip_regex = uri .. [[/?(?P<stripped_uri>.*)]]
+
+          local from, _, err = re_find(uri, [[\(.*\)]], "jo")
```

Comment:
Added it and other test cases for completeness. It gets tricky for capture groups at the start of the evaluated string, and unbalanced captured groups as well:

```lua
local uris             = {
  ["/users/\\(foo\\)"] = false,
  ["/users/\\(\\)"]    = false,

  ["/users/(foo)"]                 = true,
  ["/users/()"]                    = true,
  ["/users/()/foo"]                = true,
  ["/users/(hello(foo)world)"]     = true,
  ["/users/(hello(foo)world"]      = true,
  ["/users/(foo)/thing/(bar)"]     = true,
  ["/users/\\(foo\\)/thing/(bar)"] = true,

  -- 0-indexed capture groups
  ["()/world"]       = true,
  ["(/hello)/world"] = true,

  -- unbalanced capture groups
  ["(/hello\\)/world"] = false,
  ["/users/(foo"]      = false,
  ["/users/\\(foo)"]   = false,
  ["/users/(foo\\)"]   = false,
}

for uri, expected_to_match in pairs(uris) do
  local s = string.match(uri, "[^\\]%(.-[^\\]%)")
  s = s or string.match(uri, "^%(.-[^\\]%)")
  s = s or string.match(uri, "%(%)")
  print(uri, ":    ", s)

  if expected_to_match then
    assert(s ~= nil, uri .. " was supposed to match")

  else
    assert(s == nil, uri .. " was not supposed to match")
  end
end
```

This looks fairly complete to me now.